### PR TITLE
 [IA-3443] OUCRC web - create/update

### PR DESCRIPTION
**Pull request by tdethier**
Allows to create and update `OrgUnitChangeRequestConfigurations` through the web interface.

Related JIRA tickets : [IA-2896](https://bluesquare.atlassian.net/browse/IA-2896) (epic), [IA-3443](https://bluesquare.atlassian.net/browse/IA-3443)

## Self proofreading checklist

- [x] Did I use eslint and black formatters
- [x] Is my code clear enough and well documented
- [x] Are my typescript files well typed
- [x] New translations have been added or updated if new strings have been introduced in the frontend
- [ ] My migrations file are included
- [ ] Are there enough tests
- [ ] Documentation has been included (for new feature)

## Doc
[link](https://github.com/BLSQ/iaso/pull/1660/files#diff-390219ed0f69a73aa72e0dbcc5c38e90ce9504abeec0b942793c97fefbabf08f)

## Changes

Added two modal windows - one for preparing the creation, the other for creating/updating.

## How to test

Go to the list of  `OrgUnitChangeRequestConfigurations` , try to create a new one by clicking on the `CREATE` button.
Update an existing one by clicking on the button.

## Print screen / video

Creation:
[Screencast from 2024-09-20 16-56-57.webm](https://github.com/user-attachments/assets/6574b42e-5868-4639-934f-0ff250e77936)

## Notes

/

[IA-2896]: https://bluesquare.atlassian.net/browse/IA-2896?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[IA-3443]: https://bluesquare.atlassian.net/browse/IA-3443?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by tdethier** [comment]
@beygorghor I've just committed my latest changes. You'll find the todo list in the PR description :wink:

**Comment by beygorghor** [comment]
For the filtering of types and add some test on creation an update we created 2 other tickets:
IA-3461 and IA-3460.

https://github.com/user-attachments/assets/650e7425-a19b-4f7e-aa0d-3f5696d159e0

File: .eslintrc.json [hunk 1]
```diff
@@ -147,6 +147,7 @@
   ],
 
     "rules": {
+      "prefer-arrow-callback": "error",
```
Comment (quang-le) [inline .eslintrc.json:150]: ❤️

File: hat/assets/js/apps/Iaso/domains/orgUnits/configuration/Dialog/OrgUnitChangeRequestConfigDialog.tsx [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+                    : MESSAGES.oucrcModalCreateButton
+            }
+            onConfirm={() => handleSubmit()}
+            onCancel={() => {
+                closeDialog();
+            }}
+        >
+            {isLoadingFullConfig && <LoadingSpinner />}
+            <Typography variant="h6" component="h6">
+                Project: {config.project.name}
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/configuration/Dialog/OrgUnitChangeRequestConfigDialog.tsx]: "Project" should be translated

File: hat/assets/js/apps/Iaso/domains/orgUnits/configuration/Dialog/OrgUnitChangeRequestConfigDialog.tsx [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+            </Typography>
+            <Typography variant="h6" component="h6">
+                Org Unit Type: {config.orgUnitType.name}
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/configuration/Dialog/OrgUnitChangeRequestConfigDialog.tsx]: "Org Unit Type" should be translated

File: hat/assets/js/apps/Iaso/domains/orgUnits/configuration/Dialog/OrgUnitChangeRequestConfigDialog.tsx [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+                    keyValue="groupSetIds"
+                    onChange={onChange}
+                    value={values.groupSetIds}
+                    errors={getErrors('groupSetIds')}
+                    label={MESSAGES.groupSetIds}
+                    options={groupSetOptions}
+                />
+            )}
+            {values?.editableFields &&
+                values.editableFields.includes('possibleTypeIds') && (
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/configuration/Dialog/OrgUnitChangeRequestConfigDialog.tsx]:
```suggestion
            {values?.editableFields?.includes('possibleTypeIds') && (
```

File: hat/assets/js/apps/Iaso/domains/orgUnits/configuration/Dialog/OrgUnitChangeRequestConfigDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+                        keyValue="possibleTypeIds"
+                        onChange={onChange}
+                        value={values.possibleTypeIds}
+                        errors={getErrors('possibleTypeIds')}
+                        label={MESSAGES.possibleTypeIds}
+                        options={orgUnitTypeOptions}
+                    />
+                )}
+            {values?.editableFields &&
+                values.editableFields.includes('possibleParentTypeIds') && (
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/configuration/Dialog/OrgUnitChangeRequestConfigDialog.tsx]:
```suggestion
            {values?.editableFields?.includes('possibleParentTypeIds') && (
```

File: hat/assets/js/apps/Iaso/domains/orgUnits/configuration/Dialog/OrgUnitChangeRequestConfigDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+                        keyValue="possibleParentTypeIds"
+                        onChange={onChange}
+                        value={values.possibleParentTypeIds}
+                        errors={getErrors('possibleParentTypeIds')}
+                        label={MESSAGES.possibleParentTypeIds}
+                        options={orgUnitTypeOptions} // Warning: we should probably filter data here (only what is available in parent/child relationship)
+                    />
+                )}
+            {values?.editableFields &&
+                values.editableFields.includes('editableReferenceFormIds') && (
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/configuration/Dialog/OrgUnitChangeRequestConfigDialog.tsx]:
```suggestion
            {values?.editableFields?.includes('editableReferenceFormIds') && (
```

File: hat/assets/js/apps/Iaso/domains/orgUnits/configuration/Dialog/OrgUnitChangeRequestConfigDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+                        keyValue="editableReferenceFormIds"
+                        onChange={onChange}
+                        value={values.editableReferenceFormIds}
+                        errors={getErrors('editableReferenceFormIds')}
+                        label={MESSAGES.editableReferenceFormIds}
+                        options={formOptions}
+                    />
+                )}
+            {values?.editableFields &&
+                values.editableFields.includes('otherGroupIds') && (
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/configuration/Dialog/OrgUnitChangeRequestConfigDialog.tsx]:
```suggestion
            {values?.editableFields?.includes('otherGroupIds') && (
```

File: hat/assets/js/apps/Iaso/domains/orgUnits/configuration/Dialog/OrgUnitChangeRequestConfigDialogCreateFirstStep.tsx [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+        errors,
+        touched,
+        formatMessage,
+        messages: MESSAGES,
+    });
+
+    const { data: allProjects, isFetching: isFetchingProjects } =
+        useGetProjectsDropdownOptions(false);
+    const { data: orgUnitTypeOptions, isFetching: isFetchingOrgUnitTypes } =
+        useGetOUCRCCheckAvailabilityDropdownOptions(values.projectId);
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/configuration/Dialog/OrgUnitChangeRequestConfigDialogCreateFirstStep.tsx:91]: dat name 😂

File: hat/assets/js/apps/Iaso/domains/orgUnits/configuration/Tables/EditableFieldsCell.tsx [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+import MESSAGES from '../messages';
+import { OrgUnitChangeRequestConfigurationFull } from '../types';
+
+export const EditableFieldsCell = ({
+    row: { original },
+}: {
+    row: { original: OrgUnitChangeRequestConfigurationFull };
+}) => {
+    const { formatMessage } = useSafeIntl();
+    const editableFields = computeEditableFields(original);
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/configuration/Tables/EditableFieldsCell.tsx]: useMemo

File: hat/assets/js/apps/Iaso/domains/orgUnits/configuration/hooks/api/useSaveOrgUnitChangeRequestConfiguration.ts [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+    // This field must be cleaned because the backend accepts only some values
+    apiValues.editable_fields = cleanEditableFieldsForSaving(
+        values.editableFields,
+    );
+
+    // All the many to many fields are added if they have a value
+    const splitAndMapToNumbers = (str?: string) => {
+        return str?.trim() ? str.split(',').map(Number) : [];
+    };
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/configuration/hooks/api/useSaveOrgUnitChangeRequestConfiguration.ts]: move it out of `mapValuesForSaving` , like `cleanEditableFieldsForSaving`

File: hat/assets/js/apps/Iaso/domains/orgUnits/configuration/hooks/api/useSaveOrgUnitChangeRequestConfiguration.ts [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+    );
+    apiValues.group_set_ids = splitAndMapToNumbers(values.groupSetIds);
+    apiValues.editable_reference_form_ids = splitAndMapToNumbers(
+        values.editableReferenceFormIds,
+    );
+    apiValues.other_group_ids = splitAndMapToNumbers(values.otherGroupIds);
+    return apiValues;
+};
+
+const patchOrgUniType = async (
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/configuration/hooks/api/useSaveOrgUnitChangeRequestConfiguration.ts]: `patchOrgUnitType`?

Also, why not `patchOUCRC` or something in that vein?

File: hat/assets/js/apps/Iaso/domains/orgUnits/configuration/hooks/api/useSaveOrgUnitChangeRequestConfiguration.ts [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+    configId: number,
+    body: ApiValues,
+): Promise<any> => {
+    const url = `${apiUrlOUCRC}${configId}/`;
+    return patchRequest(url, body);
+};
+
+const postOrgUnitType = async (body: ApiValues): Promise<any> => {
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/configuration/hooks/api/useSaveOrgUnitChangeRequestConfiguration.ts]: same same

File: hat/assets/js/apps/Iaso/domains/orgUnits/configuration/hooks/useValidationSchemaOUCRC.ts [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+            .nullable()
+            .when('orgUnitsEditable', {
+                is: true,
+                then: yup.string().nullable(),
+                otherwise: yup.string().nullable(),
+            }),
+        otherGroupIds: yup.string().nullable().when('orgUnitsEditable', {
+            is: true,
+            then: yup.string().nullable(),
+            otherwise: yup.string().nullable(),
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/configuration/hooks/useValidationSchemaOUCRC.ts]: This looks the same to me as `otherGroupIds: yup.string().nullable()`

File: hat/assets/js/apps/Iaso/domains/orgUnits/configuration/hooks/useValidationSchemaOUCRC.ts [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+                then: yup.string().nullable().isMultiSelectValid(formatMessage),
+                otherwise: yup.string().nullable(),
+            }),
+        editableReferenceFormIds: yup
+            .string()
+            .nullable()
+            .when('orgUnitsEditable', {
+                is: true,
+                then: yup.string().nullable(),
+                otherwise: yup.string().nullable(),
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/configuration/hooks/useValidationSchemaOUCRC.ts]: same

File: hat/assets/js/apps/Iaso/domains/orgUnits/configuration/hooks/useValidationSchemaOUCRC.ts [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+            .nullable()
+            .required(formatMessage(MESSAGES.requiredField)),
+        orgUnitsEditable: yup
+            .boolean()
+            .nullable()
+            .required(formatMessage(MESSAGES.requiredField)),
+        editableFields: yup.string().nullable().when('orgUnitsEditable', {
+            is: true,
+            then: yup.string().nullable(),
+            otherwise: yup.string().nullable(),
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/configuration/hooks/useValidationSchemaOUCRC.ts]: same

File: iaso/api/group_sets/views.py [hunk 16]
```diff
@@ -75,3 +84,32 @@ def create(self, request, *args, **kwargs):
...
+            raise serializers.ValidationError("Parameter app_id is missing")
+
+        if user and user.is_authenticated:
+            account = user.iaso_profile.account
+            # Filter on version ids (linked to the account)
+            versions = SourceVersion.objects.filter(data_source__projects__account=account)
+
+        else:
+            # this check if project need auth
+            project = Project.objects.get_for_user_and_app_id(user, app_id)
```
Comment (quang-le) [inline iaso/api/group_sets/views.py]: This will `raise` if no project is found

**Review by quang-le — CHANGES_REQUESTED** [review]
In the code, the onmy blocking change is the one in the python code with the raised exception.

The feature works foinr for the golden path, but I can't figure out the edge cases.

On the layout side, the filter button should be aligned on the right:
Screenshot 2024-09-30 at 17 17 16

And its layout is broken on 1200 wide screens:

Screenshot 2024-09-30 at 17 17 11

Nice job, that was quite a complex layout

File: hat/assets/js/apps/Iaso/domains/orgUnits/configuration/Dialog/OrgUnitChangeRequestConfigDialog.tsx (hunk 2 above)
Comment (beygorghor) [reply hat/assets/js/apps/Iaso/domains/orgUnits/configuration/Dialog/OrgUnitChangeRequestConfigDialog.tsx:173]: c

**Review by quang-le — APPROVED** [review]

**Pull request merged**